src/main/java/org/sagebionetworks/bridge/services/ParticipantService.java

```diff
@@ -559,18 +548,16 @@ private void updateAccountAndRoles(App app, Account account, StudyParticipant pa
        
         RequestContext requestContext = RequestContext.get();
         
-        // You can no longer enroll users or add them to studies through the enrollments table just
-        // by updating the participant account. There are separate APIs for this. HOWEVER we have one 
-        // important exception: accounts that are identifiable only by an external ID. Since an external
-        // ID enrolls you in a study, administrative callers can supply study:externalId mappings 
-        // to enroll such an account at account creation.
-        if (isNew && requestContext.isAdministrator()) {
+        // New accounts can simultaneously enroll themselves in a study using an external ID.
+        // Legacy apps do this so we must continue to support it.
+        if (isNew) {
             for (Map.Entry<String, String> entry : participant.getExternalIds().entrySet()) {
                 String studyId = entry.getKey();
                 String externalId = entry.getValue();
                 
                 Enrollment enrollment = Enrollment.create(account.getAppId(), studyId, account.getId(), externalId);
-                enrollmentService.addEnrollment(account, enrollment);
+                enrollmentService.updateEnrollment(account, enrollment, enrollment);
+                account.getEnrollments().add(enrollment);
```

Comment:
EnrollmentService.addEnrollment() handles things like permissions checking and enrollment validation. Should we be bypassing those checks just to gain a bit of performance? Bypassing this layer of abstraction risks introducing bugs in the future by calling our internal API inconsistently.

Relatedly, should EnrollmentService.updateEnrollment actually be private?

Reply:
A better solution is to acquire the account ID in the request context, and then call the original method which will then succeed, so I switched to doing that, and made the updateEnrollment method private. For the most part that was fine, I did have to remove a couple of assertions in the ParticipantService test, but that test shouldn't be testing the behavior of the EnrollmentService anyway.